src/java/org/apache/cassandra/cql3/Validation.java

```diff
@@ -76,18 +78,18 @@ public static void validateKeyAndCheckConstraints(TableMetadata metadata, ByteBu
         validateKey(metadata, key);
 
         List<ColumnMetadata> partitionKeys = metadata.partitionKeyColumns();
-        List<ColumnConstraint> columnConstraints = new ArrayList<>(partitionKeys.size());
+        List<ColumnConstraint> partitionKeyConstraints = new ArrayList<>(partitionKeys.size());
         for (ColumnMetadata column : partitionKeys)
         {
-            columnConstraints.add(column.getColumnConstraints());
+            partitionKeyConstraints.add(column.getColumnConstraints());
```

Comment:
How about skipping the columns that do not have constraints? It should minimize the list and exit early when no constraint is defined for partition key, avoiding creating temporary objects.